From 2b6f4eecd26b1575e780aa0b89a943e58292933d Mon Sep 17 00:00:00 2001
From: Li Wencheng <liwencheng@phytium.com.cn>
Date: Fri, 26 Apr 2024 14:57:15 +0800
Subject: [PATCH 17/67] PHYTIUM: net/phytmac: Modify the method of obtaining
 the link status

Firstly, modify the state of pdata->speed and pdata->duplex to be
obtained in the phytmac_mac_config.
Secondly, the base.speed and base.duplex in the phytmac_get_link_
ksettings were changed from the default values to pdata.

Mainline: NA
Signed-off-by: Li Wencheng <liwencheng@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
Change-Id: Ifda2028cfd32381f97b57ad1a46f0861c0c21105

Link: https://gitee.com/phytium_embedded/phytium-linux-kernel/commit/a728bf473513902c773a61d68ea35a1dd08c39d6
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 .../net/ethernet/phytium/phytmac_ethtool.c    | 45 +++++++------------
 drivers/net/ethernet/phytium/phytmac_pci.c    |  2 +-
 drivers/net/ethernet/phytium/phytmac_v2.c     |  1 +
 3 files changed, 19 insertions(+), 29 deletions(-)

diff --git a/drivers/net/ethernet/phytium/phytmac_ethtool.c b/drivers/net/ethernet/phytium/phytmac_ethtool.c
index 592d2d9dc6d4..7cddb01802f2 100644
--- a/drivers/net/ethernet/phytium/phytmac_ethtool.c
+++ b/drivers/net/ethernet/phytium/phytmac_ethtool.c
@@ -382,43 +382,32 @@ static int phytmac_get_link_ksettings(struct net_device *ndev,
 	u32 advertising = 0;

 	if (!ndev->phydev) {
+		kset->base.port = PORT_FIBRE;
+		kset->base.transceiver = XCVR_INTERNAL;
+		kset->base.duplex = pdata->duplex;
+		kset->base.speed = pdata->speed;
+
 		if (pdata->phy_interface == PHY_INTERFACE_MODE_USXGMII ||
 		    pdata->phy_interface == PHY_INTERFACE_MODE_10GBASER) {
 			supported = SUPPORTED_10000baseT_Full
 				    | SUPPORTED_FIBRE | SUPPORTED_Pause;
 			advertising = ADVERTISED_10000baseT_Full
 				      | ADVERTISED_FIBRE | ADVERTISED_Pause;
-			kset->base.port = PORT_FIBRE;
-			kset->base.transceiver = XCVR_INTERNAL;
-			kset->base.duplex = DUPLEX_FULL;
-			kset->base.speed = SPEED_10000;
 		}  else if (pdata->phy_interface == PHY_INTERFACE_MODE_2500BASEX) {
-			supported = SUPPORTED_2500baseX_Full | SUPPORTED_Pause;
-			advertising = ADVERTISED_2500baseX_Full | ADVERTISED_Pause;
-			kset->base.port = PORT_FIBRE;
-			kset->base.transceiver = XCVR_INTERNAL;
-			kset->base.duplex = DUPLEX_FULL;
-			kset->base.speed = SPEED_2500;
+			supported = SUPPORTED_2500baseX_Full
+				    | SUPPORTED_FIBRE | SUPPORTED_Pause;
+			advertising = ADVERTISED_2500baseX_Full
+				      | ADVERTISED_FIBRE | ADVERTISED_Pause;
 		} else if (pdata->phy_interface == PHY_INTERFACE_MODE_1000BASEX) {
-			supported = SUPPORTED_1000baseT_Full | SUPPORTED_100baseT_Full
-				    | SUPPORTED_10baseT_Full | SUPPORTED_FIBRE
-				    | SUPPORTED_Pause;
-			advertising = ADVERTISED_1000baseT_Full | ADVERTISED_100baseT_Full
-				      | ADVERTISED_10baseT_Full | ADVERTISED_FIBRE
-				      | ADVERTISED_Pause;
-			kset->base.port = PORT_FIBRE;
-			kset->base.transceiver = XCVR_INTERNAL;
-			kset->base.duplex = DUPLEX_FULL;
-			kset->base.speed = SPEED_100;
+			supported = SUPPORTED_1000baseT_Full
+				    | SUPPORTED_FIBRE | SUPPORTED_Pause;
+			advertising = ADVERTISED_1000baseT_Full
+				      | ADVERTISED_FIBRE | ADVERTISED_Pause;
 		} else if (pdata->phy_interface == PHY_INTERFACE_MODE_SGMII) {
-			supported = SUPPORTED_1000baseT_Full | SUPPORTED_100baseT_Full
-				|  SUPPORTED_10baseT_Full | SUPPORTED_FIBRE | SUPPORTED_Pause;
-			advertising = ADVERTISED_1000baseT_Full | ADVERTISED_100baseT_Full
-				| ADVERTISED_10baseT_Full | ADVERTISED_FIBRE | ADVERTISED_Pause;
-			kset->base.port = PORT_FIBRE;
-			kset->base.transceiver = XCVR_INTERNAL;
-			kset->base.duplex = DUPLEX_FULL;
-			kset->base.speed = SPEED_1000;
+			supported = SUPPORTED_1000baseT_Full
+				    | SUPPORTED_FIBRE | SUPPORTED_Pause;
+			advertising = ADVERTISED_1000baseT_Full
+				      | ADVERTISED_FIBRE | ADVERTISED_Pause;
 		}

 		ethtool_convert_legacy_u32_to_link_mode(kset->link_modes.supported,
diff --git a/drivers/net/ethernet/phytium/phytmac_pci.c b/drivers/net/ethernet/phytium/phytmac_pci.c
index fd21bf80f138..af69329fe06d 100644
--- a/drivers/net/ethernet/phytium/phytmac_pci.c
+++ b/drivers/net/ethernet/phytium/phytmac_pci.c
@@ -225,7 +225,7 @@ struct phytmac_data phytmac_1000basex = {
 	.use_mii = false,
 	.speed = 1000,
 	.duplex = true,
-	.interface = PHY_INTERFACE_MODE_SGMII,
+	.interface = PHY_INTERFACE_MODE_1000BASEX,
 	.properties = fl_properties[0],
 };

diff --git a/drivers/net/ethernet/phytium/phytmac_v2.c b/drivers/net/ethernet/phytium/phytmac_v2.c
index df142aa6797f..41e5df41226e 100644
--- a/drivers/net/ethernet/phytium/phytmac_v2.c
+++ b/drivers/net/ethernet/phytium/phytmac_v2.c
@@ -842,6 +842,7 @@ static int phytmac_pcs_linkdown(struct phytmac *pdata)
 static unsigned int phytmac_pcs_get_link(struct phytmac *pdata, phy_interface_t interface)
 {
 	if (interface == PHY_INTERFACE_MODE_SGMII ||
+	    interface == PHY_INTERFACE_MODE_1000BASEX ||
 	    interface == PHY_INTERFACE_MODE_2500BASEX)
 		return PHYTMAC_READ_BITS(pdata, PHYTMAC_NETWORK_STATUS, LINK);
 	else if (interface == PHY_INTERFACE_MODE_USXGMII ||
--
2.50.1
